src/rds_scheduler.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+require 'bundler'
+Bundler.setup
+
+require 'aws-sdk-rds'
+require 'logger'
+
+require_relative 'rds_helper'
+require_relative 'time_parser'
+
+# Manage uptime schedules for RDS Instances
+class RDSScheduler
+  attr_reader :loop_interval,
+              :run_once,
+              :tag_uptime_schedule
+
+  def initialize
+    @logger = Logger.new(STDOUT)
+    $stdout.sync = true
+    dry_run = ENV.fetch('DRY_RUN', false).to_s.casecmp('true').zero?
+    @loop_interval = ENV.fetch('LOOP_INTERVAL', 30).to_i
+    @rds_client = RDSHelper.new(dry_run: dry_run, logger: @logger)
+    @run_once = ENV.fetch('RUN_ONCE', false).to_s.casecmp('true').zero?
+    @tag_uptime_schedule = ENV.fetch('TAG_UPTIME_SCHEDULE', 'appvia.io/rds-scheduler/uptime-schedule')
+    @time_parser = TimeParser.new
+  end
+
+  def execute
+    loop do
+      @logger.info 'Retrieving DB instances...'
+      dbs = @rds_client.db_instances
+
+      dbs.each do |rds|
+        tags = @rds_client.db_tags(rds.db_instance_arn)
+
+        db_schedule = false
+        tags.each do |tag|
+          (db_schedule = tag.value) && break if tag.key.eql?(tag_uptime_schedule)
+        end
+
+        process_schedule(rds.db_instance_identifier, rds.db_instance_status, db_schedule)
+      end
+
+      break if run_once
+
+      @logger.info("Sleeping for #{loop_interval} seconds...")
+      sleep(loop_interval)
+    end
+  end
+
+  def process_schedule(db_name, db_status, db_schedule)
+    if db_schedule
+      begin
+        parsed_schedule = @time_parser.parse_schedule(db_schedule)
+      rescue StandardError => e
+        @logger.warn "DB Instance '#{db_name}' has an invalid schedule: #{e.message}"
+      else
+        if @time_parser.schedule_matches?(parsed_schedule)
+          @rds_client.start_db_instance(db_name, db_status, db_schedule)
+        else
+          @rds_client.stop_db_instance(db_name, db_status, db_schedule)
```

Comment:
I wonder if your main script needs to catch `INT` and `TERM` signals and handle graceful shutdown ([very simple example](https://calendar.google.com)); are these start and stop operations blocking or async? In either case, what happens if the pod etc. is sent a `TERM` signal whilst they are going on?

Reply:
It's an API call to RDS to initiate starting / stopping the instance, gets an immediate response back. No issues to RDS if it is interrupted.